src/common/plist.mligo

```diff
@@ -0,0 +1,98 @@
+// SPDX-FileCopyrightText: 2021 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+#if !COMMON_PLIST_H
+#define COMMON_PLIST_H
+
+#include "../types.mligo"
+
+[@inline]
+let plist_mem (key, plist_o : proposal_key * proposal_doubly_linked_list option): bool =
+  match plist_o with
+    | None -> false
+    | Some plist ->
+        if (key = plist.first)
+          then true
+          else Map.mem (key, prev) plist.map
+
+[@inline]
+let plist_pop (plist_o : proposal_doubly_linked_list option): proposal_doubly_linked_list option =
+  match plist_o with
+    | None -> None
+    | Some plist ->
+        (match Map.find_opt (plist.first, next) plist.map with
+            | Some next_key ->
+                Some
+                  { plist with
+                    first = next_key // Make the next as the first key
+                  ; map =
+                      Map.remove (next_key, prev) // Remove the previous link of the next key.
+                        (Map.remove (plist.first, next) plist.map) // Remove the first key link.
+                  }
+            | None -> (None : proposal_doubly_linked_list option) // Only the first key exist.
+        )
```

Comment:
This definitely looks more like what I had in mind and I don't have any comment on the logic (or its efficiency), however I think that it should be changed to return the first key itself (if any) as well as the updated list:
```suggestion
let plist_pop (plist_o : proposal_doubly_linked_list option)
    : (proposal_key option * proposal_doubly_linked_list option) =
  match plist_o with
  | None ->
    ( (None : proposal_key option)
    , (None : proposal_doubly_linked_list option)
    )
  | Some plist ->
    let new_plist = match Map.find_opt (plist.first, next) plist.map with
      | Some next_key ->
        Some
          { plist with
            first = next_key // Make the next as the first key
          ; map =
              Map.remove (next_key, prev) // Remove the previous link of the next key.
                (Map.remove (plist.first, next) plist.map) // Remove the first key link.
          }
      | None -> (None : proposal_doubly_linked_list option) // Only the first key exist.
    in ( Some plist.first, new_plist)
```
because it would save more work in its `proposal.mligo` use (see next comment).

Reply:
I was thinking that since we can access first element easily, we don't need to return it, but as you mentioned below, it definitely helps in our use case. I've updated this.